unit_scaling/transforms/_unit_scale.py

```diff
@@ -0,0 +1,294 @@
+            unit_scaling_backend_idx = i
+        if "quantisation_backend" in b.__qualname__:
+            quantisation_backend_idx = i
+    if unit_scaling_backend_idx > quantisation_backend_idx:
+        logger.info("moving unit scaling backend to precede quantisation backend")
+        u = backends.pop(unit_scaling_backend_idx)
+        backends.insert(quantisation_backend_idx, u)  # type: ignore[arg-type]
+
+
+M = TypeVar("M", bound=nn.Module)
+
+
+def unit_scale(
+    module: M, replace: Dict[Callable[..., Any], Callable[..., Any]] = dict()
+) -> M:
+    """**[Experimental]** Returns a unit-scaled version of the input model.
+
+    Uses TorchDynamo to trace and transform the user-supplied module.
+    This transformation identifies all :mod:`torch.nn.functional` uses in the input
+    module, and replaces them with their unit-scaled equivalents, should they exist.
+
+    The tracing procedure automatically recurses into modules
+    (whether defined in libraries, or by the user), identifying inner calls to any
+    :mod:`torch.nn.functional` operations, to build a graph of fundamental operations.
+    Unit scaling is then applied as a transformation on top of this graph.
+
+    This transformation proceeds in five stages:
+
+    #. **Replacement of user-defined functions** according to the supplied `replace`
+       dictionary.
+    #. **Replacement of all functions with unit-scaled equivalents** defined in
+       :mod:`unit_scaling.functional`.
+    #. Identification & **replacement of all add operations that represent
+       residual-adds**. The identification of residual connections is done via a
+       dependency analysis on the graph. Residual-adds require special scaling compared
+       with regular adds (see paper / User Guide for details).
+    #. **Unconstraining of all operations after the final residual layer**. By default
+       all unit scaled operations have their scaling factors constrained in the forward
+       and backward pass to give valid gradients. This is not required in these final
+       layers (see paper for proof), and hence we can unconstrain the operations to give
+       better scaling.
+    #. **Unit-scaling of all weights** and zero-initialisation of all biases.
+
+    Note that by using TorchDynamo, `unit_scale()` is able to trace a much larger set of
+    modules / operations than with previous PyTorch tracing approaches. This enables
+    the process of unit scaling to be expressed as a generic graph transform that can be
+    applied to arbitrary modules.
+
+    Note that the current version of TorchDynamo (or :func:`torch.compile`, which is a
+    wrapper around TorchDynamo) doesn't support nested transforms, so we implement our
+    own system here. This makes it easy to nest transforms:
+
+    .. code-block:: python
+
+        from unit_scaling.transforms import compile, simulate_fp8, unit_scale
+
+        module = compile(simulate_fp8(unit_scale(module)))
```

Comment:
I can't see where `unit_scaling.transforms.compile` is defined..?

Reply:
This docstring was written in anticipation of me implementing it 🙈